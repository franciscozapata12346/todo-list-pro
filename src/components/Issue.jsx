import React, { useState } from 'react';

export default function Issue({id, done, name, updateIssue, deleteIssue}) {

    const [editMode, setEditMode] = useState(false)

    // Marca una tarea como finalizda o pendiente
    const markAsDone = () => {
        updateIssue({
            id,
            name,
            done: !done
        })
    }

    // Permite cambiar el nombre de la issue
    const enableEditMode = (e) => {
        e.stopPropagation()
        setEditMode(true)
    }

    // Cambia el nombre de la issue
    const editName = (e) => {
        e.stopPropagation()

        if(e.type === 'keydown' && e.key !== 'Enter') {
            return 
        }
        const issueName = e.target.value.trim()

        if(issueName.length === 0) return

        updateIssue({
            id,
            name: issueName,
            done
        })
        setEditMode(false)
    }

    // Elimina una issue de la lista
    const handleDeleteIssue = (e) => {
        e.stopPropagation()
        deleteIssue(id)
    }

    return (
        <div className="issue" 
             onClick={() => markAsDone(!done) }>

            {!editMode &&
            <span className={"issue__name " + 
                            (done ? "issue__name--done" : "")}>
                {name}
            </span>}

            {editMode &&
            <input type="text" 
                   className="issue__input" 
                   defaultValue={name} 
                   autoFocus
                   onKeyDown={editName}
                   onBlur={editName}/>
            }

            <div className="issue__actions-box">
                {!editMode && 
                <span className="issue__icon-button issue__icon-button--edit"
                      onClick={enableEditMode}>
                    <i className="fa fa-pen"></i>
                </span>}
                <span className="issue__icon-button issue__icon-button--delete"
                      onClick={handleDeleteIssue}>
                    <i className="fa fa-trash"></i>
                </span>
            </div>
        </div>
    )
}
