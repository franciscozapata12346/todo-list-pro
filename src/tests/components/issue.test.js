import { shallow } from "enzyme"
import Issue from "../../components/Issue"

describe('Pruebas a <Issue />', () => {
    
    const id = Date.now()
    const name = 'Sacar basura'
    const done = false
    const setIssues = jest.fn()

    let wrapper = shallow(<Issue id={id} name={name} done={done} setIssues={setIssues} />)

    beforeEach(() => {
        jest.clearAllMocks()
        wrapper = shallow(<Issue id={id} name={name} done={done} setIssues={setIssues} />)
    })
    

    test('El componente debe mostrarse correctamente', () => {
        
        expect(wrapper).toMatchSnapshot()
    })
    

    test('El texto de la issue debe coincidir con el prop', () => {

        const issueName = wrapper.find('.issue__name').text().trim()
        expect(issueName).toBe(name)
    })


    test('Cuando done = true debe aplicar la clase --done', () => {
        
        const wrapper = shallow(<Issue id={id} name={name} done={true} setIssues={setIssues}/>)
        const className = wrapper.find('.issue__name').prop('className')
        expect(className.includes('--done')).toBeTruthy()
    })
    
    
    test('Cuando done = false no debe aplicar la clase --done', () => {  

        const className = wrapper.find('.issue__name').prop('className')
        expect(className.includes('--done')).toBeFalsy()
    })


    test('Cuando hacemos click editar debe aparecer el input y desaparecer span', () => {

        const editButton = wrapper.find('.issue__icon-button--edit')
        editButton.simulate('click', { stopPropagation(){}})

        expect(wrapper.find('.issue__input').exists()).toBe(true)
        expect(wrapper.find('.issue__name').exists()).toBe(false)
    })


    test('Eliminar la issue debe llamar a setIssues una vez', () => {

        const deleteButton = wrapper.find('.issue__icon-button--delete')
        deleteButton.simulate('click', { stopPropagation(){}})

        expect(setIssues).toBeCalled()
    })


    test('Modificar la issue debe llamar a setIssues una vez', () => {

        const editButton = wrapper.find('.issue__icon-button--edit')
        editButton.simulate('click', { stopPropagation(){}})

        const input = wrapper.find('.issue__input')
        const value = 'Amasar la masa'
        input.simulate('change', { target: {value} })
        input.simulate('blur', { stopPropagation(){}, target: {value} })
        
        expect(setIssues).toBeCalled()
    })

    
})
