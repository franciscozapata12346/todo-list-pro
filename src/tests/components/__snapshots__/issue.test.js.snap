// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas a <Issue /> El componente debe mostrarse correctamente 1`] = `
<div
  className="issue"
  onClick={[Function]}
>
  <span
    className="issue__name "
  >
    Sacar basura
  </span>
  <div
    className="issue__actions-box"
  >
    <span
      className="issue__icon-button issue__icon-button--edit"
      onClick={[Function]}
    >
      <i
        className="fa fa-pen"
      />
    </span>
    <span
      className="issue__icon-button issue__icon-button--delete"
      onClick={[Function]}
    >
      <i
        className="fa fa-trash"
      />
    </span>
  </div>
</div>
`;
